onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -507,26 +514,48 @@ class PlannerImpl {
       auto process_input = [&graph_inputs, &exec_provider, &p_kernel_def, &is_implicit_input,
+                            &node_arg_has_explicit_consumer,
                             this](const NodeArg& input, size_t arg_idx) {
         const auto& name = input.Name();
         UseCount(name)++;
 
+        bool is_graph_input = (graph_inputs.find(name) != graph_inputs.cend());
+        bool is_outer_scope_arg = std::find_if(outer_scope_node_args_.cbegin(), outer_scope_node_args_.cend(),
+                                               [&name](const NodeArg* value) {
+                                                 return value && value->Name() == name;
+                                               }) != outer_scope_node_args_.cend();
+        bool is_subgraph = (parent_node_ != nullptr);
+
         // If it's a graph input or outer scope node arg, set its plan.
         // NOTE: Copy nodes should have already been added if a graph input is fed as input
         // to nodes assigned to different providers.
-        if (graph_inputs.find(name) != graph_inputs.cend() ||
-            std::find_if(outer_scope_node_args_.cbegin(), outer_scope_node_args_.cend(),
-                         [&name](const NodeArg* value) {
-                           return value && value->Name() == name;
-                         }) != outer_scope_node_args_.cend()) {
+
+        if (is_graph_input || is_outer_scope_arg) {
           OrtValueIndex index = Index(name);
 
-          // implicit inputs do not have an entry in the kernel def, so do nothing to them here, leaving the control
-          //   flow op (Loop, Scan, If) to do the necessary copy if the input crosses different provider.
-          // matching logic is used in TransformerMemcpyImpl::ProcessDefs
           if (!is_implicit_input) {
             OrtMemType mem_type = p_kernel_def->InputMemoryType(arg_idx);
             plan_.SetLocation(static_cast<size_t>(index), exec_provider->GetAllocator(0, mem_type)->Info());
+          } else {  // implicit input
+            // Only process an implicit input:
+            // 1) Within a subgraph
+            // 2) If it is a non-subgraph input (i.e.) only an outer scope node arg fed through
+            // to this level of the subgraph as an implicit input
+            // 3) If there is no explicit consumer at this graph level
+            // If there is an explicit consumer, we want the location to reflect that
+            // and not where it is produced in the outer scope.
+            // It is okay if we process a node consuming this arg as an implicit input
+            // ahead of a node that is an explicit consumer, because we will just reset
+            // this location in the 'if' branch above.
+
+            // TODO: Can we extend this optimization to explicit subgraph inputs that are
+            // "passed through" to a nested subgraph (i.e.) when there is no explicit
+            // consumer of a subgraph input at this level of the subgraph.
```

Comment:
Maybe say we 'can', but call out the details like you need to handle the name change between the input to the control flow node and the explicit subgraph input, and that there's some per-operator logic required to map between those two things. That insight is non-trivial and important to understanding what that optimization would need to do.

Reply:
I have left this comment in session state where we "accumulate" all the outer scope locations to be provided to the subgraph planner. I feel it is more "relevant" there as that is where we need to do the mapping. We don't need any logic in the allocation planner (such information should already be populated in the map by the time control reaches here).

The more I think about it, the only additional logic is the "name change", not sure if we really do need operator specific logic. "If" doesn't have explicit subgraph inputs. Loop and Scan have explicit subgraph inputs that are the same as the operator's inputs. So all we really need to do is execute the extra logic when explicit subgraph inputs > 0, I would have thought. That isn't "operator specific" is it ?

Reply:
yeah - if you're just tracking the new names it should be generic.

maybe link this with the fuller comment by saying please see OuterScopeNodeArgLocationAccumulator in session_state.cc

Reply:
Removed all comments pertaining to extending this optimization for explicit subgraph inputs and added the logic here as it gives huge gains for the second model.